onnxruntime/core/providers/cpu/rnn/rnn_helpers.h

```diff
@@ -230,17 +230,49 @@ void ExecuteLambdaInParallel(const std::string& name, TLambda lambda, int max, i
   ORT_UNUSED_PARAMETER(name);
   ORT_UNUSED_PARAMETER(logger);
 
-  std::atomic<int> done(0);
-  for (int i = 0; i < max; i += step) {
-    ttp.Schedule([lambda, i, &done]() {
-      lambda(i);
-      ++done;
+  // ORT_ENFORCE may and does throw at times from within the tasks that run
+  // on a thread-pool. Without propagating exceptions the process exits silently
+  // which will make diagnosing bugs more difficult.
+
+  // \! UGLY
+  // We have a problem here with the current thread-pool is that it takes std::function
+  // by value and copies it more than once (even though it is movable).
+  //
+  // To report status and exceptions properly it's better to use
+  // futures and promises but they are not copyable, so we can't come up with a functor
+  // with a promise member and we are downgrading to C++11 where we can't have captures that moved in.
+  //
+  // At the same time promises MUST live in the child thread so if we throw from the main thread
+  // we don't destroy any promises that are on the main thread stack which children threads may still be using.
+  //
+  // The only solution with the current Eigen that comes to mind is to have shared_ptr to with std::promise.
+  //
+
+  std::vector<std::future<void> > futures;
+  for (int i = 0, t = 0; i < max; i += step, ++t) {
+    auto p_ptr = std::make_shared<std::promise<void> >();
+    futures.push_back(p_ptr->get_future());
+    ttp.Schedule([p_ptr, lambda, i]() {
+      try {
+        lambda(i);
+        p_ptr->set_value();
+      } catch (...) {
+        p_ptr->set_exception(std::current_exception());
+      }
     });
   }
 
-  int totalTasks = max / (step > 0 ? step : 1) + (max % step > 0 ? 1 : 0);
-  while (done != totalTasks)
-    ;
+  // If ORT_ENFORCE is failing multiple threads can throw
+  // we ignore other exceptions but one. Futures with unclaimed
+  // shared states should destruct just fine.
+  for (auto& fut : futures) {
+    try {
+      fut.get();
+    } catch (...) {
+      throw;
+    }
```

Comment:
What's the purpose of the catch and rethrow vs. not catching?

Reply:
Good point.

---
In reply to: [327852017](https://github.com/microsoft/onnxruntime/pull/1905#discussion_r327852017) [](ancestors = 327852017)

Reply:
I think it should catch all the exceptions and only throw the first one here. It must wait for all the subtasks finish.

Reply:
Yeah, I thought the get() was meant for waiting.

Reply:
Sounds good, that may prevent crashes should the stack unwinding destroys the actual kernel on which other tasks are still running.

---
In reply to: [327855620](https://github.com/microsoft/onnxruntime/pull/1905#discussion_r327855620) [](ancestors = 327855620)

Reply:
Currently, we wait until the first throw. @snnn proposes to store the first exception and not to re-throw untill all are done.

---
In reply to: [327860672](https://github.com/microsoft/onnxruntime/pull/1905#discussion_r327860672) [](ancestors = 327860672)